README.md
# noticeable

Noticeable color differences


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.1.2 (2021-03-16)


src/lib/step.spec.ts
import test from 'ava';
import * as d3 from 'd3-color';

import { step } from './step';

test('max lab.l', (t) => {
  const lab = d3.lab('white');
  t.is(lab.l, 100);
});

test('white', (t) => {
  const lab = d3.lab('white');
  const labMod = d3.lab(lab.l - step.l, lab.a, lab.b);
  t.is(labMod.toString(), 'rgb(225, 225, 225)');
});

test('black', (t) => {
  const lab = d3.lab('black');
  const labMod = d3.lab(lab.l + step.l, lab.a, lab.b);
  t.is(labMod.toString(), 'rgb(29, 29, 29)');
});

test('grey', (t) => {
  const lab = d3.lab('grey');
  const lighter = d3.lab(lab.l + step.l, lab.a, lab.b);
  t.is(lighter.toString(), 'rgb(155, 155, 155)');
  const darker = d3.lab(lab.l - step.l, lab.a, lab.b);
  t.is(darker.toString(), 'rgb(102, 102, 102)');
});

// test('reddish', (t) => {
//   const lab = d3.lab('rgb(120, 90, 60)');
//   const lighter = d3.lab(lab.l + step.l, lab.a, lab.b);
//   t.is(lighter.toString(), 'rgb()');
//   const darker = d3.lab(lab.l - step.l, lab.a, lab.b);
//   t.is(darker.toString(), 'rgb()');
// });


assets/js/search.js
window.searchData = {"kinds":{"1":"Module","32":"Variable","1024":"Property","65536":"Type literal"},"rows":[{"id":0,"kind":1,"name":"index","url":"modules/index.html","classes":"tsd-kind-module"},{"id":1,"kind":1,"name":"lib/step","url":"modules/lib_step.html","classes":"tsd-kind-module"},{"id":2,"kind":32,"name":"step","url":"modules/lib_step.html#step","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"lib/step"},{"id":3,"kind":65536,"name":"__type","url":"modules/lib_step.html#step.__type","classes":"tsd-kind-type-literal tsd-parent-kind-variable","parent":"lib/step.step"},{"id":4,"kind":1024,"name":"l","url":"modules/lib_step.html#step.__type.l","classes":"tsd-kind-property tsd-parent-kind-type-literal","parent":"lib/step.step.__type"},{"id":5,"kind":1024,"name":"a","url":"modules/lib_step.html#step.__type.a","classes":"tsd-kind-property tsd-parent-kind-type-literal","parent":"lib/step.step.__type"},{"id":6,"kind":1024,"name":"b","url":"modules/lib_step.html#step.__type.b","classes":"tsd-kind-property tsd-parent-kind-type-literal","parent":"lib/step.step.__type"},{"id":7,"kind":32,"name":"whiter","url":"modules/lib_step.html#whiter","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"lib/step"},{"id":8,"kind":32,"name":"blacker","url":"modules/lib_step.html#blacker","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"lib/step"},{"id":9,"kind":32,"name":"redder","url":"modules/lib_step.html#redder","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"lib/step"},{"id":10,"kind":32,"name":"greener","url":"modules/lib_step.html#greener","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"lib/step"},{"id":11,"kind":32,"name":"yellower","url":"modules/lib_step.html#yellower","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"lib/step"},{"id":12,"kind":32,"name":"bluer","url":"modules/lib_step.html#bluer","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"lib/step"},{"id":13,"kind":32,"name":"step","url":"modules/index.html#step","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"index"},{"id":14,"kind":65536,"name":"__type","url":"modules/index.html#step.__type","classes":"tsd-kind-type-literal tsd-parent-kind-variable","parent":"index.step"},{"id":15,"kind":1024,"name":"l","url":"modules/index.html#step.__type.l","classes":"tsd-kind-property tsd-parent-kind-type-literal","parent":"index.step.__type"},{"id":16,"kind":1024,"name":"a","url":"modules/index.html#step.__type.a","classes":"tsd-kind-property tsd-parent-kind-type-literal","parent":"index.step.__type"},{"id":17,"kind":1024,"name":"b","url":"modules/index.html#step.__type.b","classes":"tsd-kind-property tsd-parent-kind-type-literal","parent":"index.step.__type"},{"id":18,"kind":32,"name":"whiter","url":"modules/index.html#whiter","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"index"},{"id":19,"kind":32,"name":"blacker","url":"modules/index.html#blacker","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"index"},{"id":20,"kind":32,"name":"redder","url":"modules/index.html#redder","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"index"},{"id":21,"kind":32,"name":"greener","url":"modules/index.html#greener","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"index"},{"id":22,"kind":32,"name":"yellower","url":"modules/index.html#yellower","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"index"},{"id":23,"kind":32,"name":"bluer","url":"modules/index.html#bluer","classes":"tsd-kind-variable tsd-parent-kind-module","parent":"index"}],"index":{"version":"2.3.9","fields":["name","parent"],"fieldVectors":[["name/0",[0,10.788]],["parent/0",[]],["name/1",[1,10.788]],["parent/1",[]],["name/2",[2,23.026]],["parent/2",[1,1.04]],["name/3",[3,23.026]],["parent/3",[4,2.713]],["name/4",[5,23.026]],["parent/4",[6,1.896]],["name/5",[7,23.026]],["parent/5",[6,1.896]],["name/6",[8,23.026]],["parent/6",[6,1.896]],["name/7",[9,23.026]],["parent/7",[1,1.04]],["name/8",[10,23.026]],["parent/8",[1,1.04]],["name/9",[11,23.026]],["parent/9",[1,1.04]],["name/10",[12,23.026]],["parent/10",[1,1.04]],["name/11",[13,23.026]],["parent/11",[1,1.04]],["name/12",[14,23.026]],["parent/12",[1,1.04]],["name/13",[2,23.026]],["parent/13",[0,1.04]],["name/14",[3,23.026]],["parent/14",[15,2.713]],["name/15",[5,23.026]],["parent/15",[16,1.896]],["name/16",[7,23.026]],["parent/16",[16,1.896]],["name/17",[8,23.026]],["parent/17",[16,1.896]],["name/18",[9,23.026]],["parent/18",[0,1.04]],["name/19",[10,23.026]],["parent/19",[0,1.04]],["name/20",[11,23.026]],["parent/20",[0,1.04]],["name/21",[12,23.026]],["parent/21",[0,1.04]],["name/22",[13,23.026]],["parent/22",[0,1.04]],["name/23",[14,23.026]],["parent/23",[0,1.04]]],"invertedIndex":[["__type",{"_index":3,"name":{"3":{},"14":{}},"parent":{}}],["a",{"_index":7,"name":{"5":{},"16":{}},"parent":{}}],["b",{"_index":8,"name":{"6":{},"17":{}},"parent":{}}],["blacker",{"_index":10,"name":{"8":{},"19":{}},"parent":{}}],["bluer",{"_index":14,"name":{"12":{},"23":{}},"parent":{}}],["greener",{"_index":12,"name":{"10":{},"21":{}},"parent":{}}],["index",{"_index":0,"name":{"0":{}},"parent":{"13":{},"18":{},"19":{},"20":{},"21":{},"22":{},"23":{}}}],["index.step",{"_index":15,"name":{},"parent":{"14":{}}}],["index.step.__type",{"_index":16,"name":{},"parent":{"15":{},"16":{},"17":{}}}],["l",{"_index":5,"name":{"4":{},"15":{}},"parent":{}}],["lib/step",{"_index":1,"name":{"1":{}},"parent":{"2":{},"7":{},"8":{},"9":{},"10":{},"11":{},"12":{}}}],["lib/step.step",{"_index":4,"name":{},"parent":{"3":{}}}],["lib/step.step.__type",{"_index":6,"name":{},"parent":{"4":{},"5":{},"6":{}}}],["redder",{"_index":11,"name":{"9":{},"20":{}},"parent":{}}],["step",{"_index":2,"name":{"2":{},"13":{}},"parent":{}}],["whiter",{"_index":9,"name":{"7":{},"18":{}},"parent":{}}],["yellower",{"_index":13,"name":{"11":{},"22":{}},"parent":{}}]],"pipeline":[]}}

src/lib/step.ts
// See https://research.tableau.com/sites/default/files/2014CIC_48_Stone_v3.pdf

const lAxis = { A: 10.16, B: 1.5 };
const aAxis = { A: 10.68, B: 3.08 };
const bAxis = { A: 10.7, B: 5.74 };

const p = 0.8; // Eighty percent of people
const s = 0.5; // 0.5 degree (size)

const l = p * (lAxis.A + lAxis.B / s);
const a = p * (aAxis.A + aAxis.B / s);
const b = p * (bAxis.A + bAxis.B / s);

/**
 * Noticeable color differences in Lab color space
 *
 * ### Example (es module)
 * ```js
 * import { step } from 'noticeable'
 * import { lab } from 'd3-color'
 *
 * const color = d3.lab("rgb(120, 90, 60)")
 * const lighter = d3.lab(color.l + step.l, color.a, color.b)
 * console.log(lighterColor.toString)
 * // => rgb()
 * ```
 */

export const step = {
  l,
  a,
  b,
};

export const whiter = +l;
export const blacker = -l;
export const redder = +a;
export const greener = -a;
export const yellower = +b;
export const bluer = -b;
